File: uptime.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/sysctl.h>
#include <sys/time.h>
#include <time.h>
#include <unistd.h>

int main(void)
{
    struct timeval boottime;
    size_t len = sizeof(boottime);
    int mib[2] = {CTL_KERN, KERN_BOOTTIME}; // MIB for kern.boottime

    // Retrieve the boottime
    if (sysctl(mib, 2, &boottime, &len, NULL, 0) < 0)
    {
        perror("sysctl");
        exit(EXIT_FAILURE);
    }

    time_t now;
    time(&now);                                    // Get the current time
    time_t uptime_seconds = now - boottime.tv_sec; // Calculate uptime in seconds

    printf("%ld\n", uptime_seconds);

    return EXIT_SUCCESS;
}

File: uptime.rs
extern crate libc;

use libc::{sysctlbyname, timeval};
use std::mem::zeroed;
use std::ptr;
use std::time::{SystemTime, UNIX_EPOCH};

fn main() {
    unsafe {
        let mut boottime: timeval = zeroed();
        let mut len: libc::size_t = std::mem::size_of::<timeval>() as libc::size_t;
        let name = std::ffi::CString::new("kern.boottime").unwrap();

        if sysctlbyname(
            name.as_ptr(),
            &mut boottime as *mut _ as *mut _,
            &mut len,
            ptr::null_mut(),
            0,
        ) == 0
        {
            let now = SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .expect("Time went backwards")
                .as_secs() as i64;

            let uptime_seconds = now - boottime.tv_sec;
            println!("Uptime: {} seconds", uptime_seconds);
        } else {
            eprintln!("Failed to get system uptime");
        }
    }
}
